x-pack/packages/kbn-elastic-assistant/impl/assistant/api.tsx

```diff
@@ -78,7 +79,8 @@ export const fetchConnectorExecuteAction = async ({
 
     if (data.choices && data.choices.length > 0 && data.choices[0].message.content) {
       const result = data.choices[0].message.content.trim();
-      return result;
+
+      return assistantLangChain ? getFormattedMessageContent(result) : result;
```

Comment:
> A client side change was required to the assistant, because the response returned from the agent executor is JSON. 

Thoughts on doing this server side as to not leak the agent executor abstraction to the client? Or was there a specific reason this needed to be pushed to the client?

Reply:
> > A client side change was required to the assistant, because the response returned from the agent executor is JSON.
> 
> Thoughts on doing this server side as to not leak the agent executor abstraction to the client? Or was there a specific reason this needed to be pushed to the client?

In summary of our offline discussion:

- Yes, it's possible to do this server side by introducing an additional `JSON.parse` of the response from the LLM
- In short, the server side code in `x-pack/plugins/elastic_assistant/server/lib/langchain/execute_custom_llm_chain/index.ts` would look something like the following (psudocode):

```typescript
  await executor.call({ input: latestMessage[0].content });

  const rawData = llm.getActionResultData(); // the response from the actions framework

  if (rawData.choices && rawData.choices.length > 0 && rawData.choices[0].message.content) {
    const result = rawData.choices[0].message.content.trim();

    const formatted = getFormattedMessageContent(result);
    const data = convertToRecord(formatted);

    return {
      connector_id: connectorId,
      data,
      status: 'ok',
    };
  } else {
    throw new Error('Unexpected raw response from the LLM');
  }
```
- The psudocode above is very similar to the client side code in `x-pack/packages/kbn-elastic-assistant/impl/assistant/api.tsx`
- If server-side parsing is implemented, the existing client side code should be extracted into a reusable function (exported by the package), for reuse on the server

In summary, it's possible, but given the above adds an additional server side JSON parse of the LLM response and requires some additional client side refactoring, we'll reconsider the above post FF.

Reply:
Thank you for taking the time to discuss offline and for summarizing here 🙏 , path forward sounds good to me 👍